common/lib/pe.c

```diff
@@ -0,0 +1,383 @@
+    IMAGE_NT_HEADERS64 *nt_hdrs = (IMAGE_NT_HEADERS64 *)(image + dos_hdr->e_lfanew);
+    IMAGE_SECTION_HEADER *sections = (IMAGE_SECTION_HEADER *)((uintptr_t)&nt_hdrs->OptionalHeader + nt_hdrs->FileHeader.SizeOfOptionalHeader);
+
+    bool is_reloc = true;
+
+    if (nt_hdrs->FileHeader.Characteristics & IMAGE_FILE_RELOCS_STRIPPED) {
+        is_reloc = false;
+    }
+
+    if (_is_reloc) {
+        *_is_reloc = is_reloc;
+    }
+
+    uint64_t image_base = nt_hdrs->OptionalHeader.ImageBase;
+    uint64_t image_size = nt_hdrs->OptionalHeader.SizeOfImage;
+    uint64_t alignment = nt_hdrs->OptionalHeader.SectionAlignment;
+
+    bool lower_to_higher = false;
+
+    if (image_base < FIXED_HIGHER_HALF_OFFSET_64) {
+        if (!is_reloc) {
+            panic(true, "pe: Lower half images are not allowed");
+        }
+
+        lower_to_higher = true;
+    }
+
+    uint64_t slide = 0;
+    size_t try_count = 0;
+    size_t max_simulated_tries = 0x10000;
+
+    if (lower_to_higher) {
+        slide = FIXED_HIGHER_HALF_OFFSET_64 - image_base;
+    }
+
+    *physical_base = (uintptr_t)ext_mem_alloc_type_aligned(image_size, alloc_type, alignment);
+    *virtual_base = image_base;
+
+    memcpy((void *)*physical_base, image, nt_hdrs->OptionalHeader.SizeOfHeaders);
+
+    if (_image_size) {
+        *_image_size = image_size;
+    }
+
+    if (is_reloc && kaslr) {
+again:
+        slide = (rand32() & ~(alignment - 1)) + (lower_to_higher ? FIXED_HIGHER_HALF_OFFSET_64 - image_base : 0);
+
+        if (*virtual_base + slide + image_size < 0xffffffff80000000 /* this comparison relies on overflow */) {
+            if (++try_count == max_simulated_tries) {
+                panic(true, "pe: Image wants to load too high");
+            }
+            goto again;
+        }
+    }
+
+    for (size_t i = 0; i < nt_hdrs->FileHeader.NumberOfSections; i++) {
+        IMAGE_SECTION_HEADER *section = &sections[i];
+
+        uint64_t section_base = *physical_base + section->VirtualAddress;
+        uint64_t section_raw_size = section->VirtualSize < section->SizeOfRawData ? section->VirtualSize : section->SizeOfRawData;
+
+        memcpy((void *)section_base, image + section->PointerToRawData, section_raw_size);
+    }
+
+    IMAGE_DATA_DIRECTORY *import_dir = &nt_hdrs->OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT];
+    IMAGE_DATA_DIRECTORY *reloc_dir = &nt_hdrs->OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_BASERELOC];
+
+    if (import_dir->Size != 0) {
+        IMAGE_IMPORT_DESCRIPTOR *import_desc = (IMAGE_IMPORT_DESCRIPTOR *)(*physical_base + import_dir->VirtualAddress);
+
+        if (import_desc->Name != 0) {
+            panic(true, "pe: Kernel must not have any imports");
+        }
+    }
+
+    if (reloc_dir->VirtualAddress != 0) {
+        IMAGE_BASE_RELOCATION_BLOCK *block = (IMAGE_BASE_RELOCATION_BLOCK *)(*physical_base + reloc_dir->VirtualAddress);
+
+        while (block->VirtualAddress != 0) {
```

Comment:
don't think this is guaranteed, you should instead use reloc_dir->size and check the offset you are into the section.